promdump/promdump.go

```diff
@@ -368,7 +381,113 @@ func exportMetric(ctx context.Context, promApi v1.API, metric string, beginTS ti
 			}
 		}
 	}
-	return writeFile(&values, filePrefix, fileNum)
+
+	return fileNum, writeFile(&values, filePrefix, fileNum)
+
+}
+
+func createArchive(buf io.Writer) error {
+	// Create new Writers for gzip , bzip2 and tar
+	// These writers are chained. Writing to the tar writer will
+
+	var tw *tar.Writer
+
+	switch *compression {
+	case "gzip":
+		gw := gzip.NewWriter(buf)
+		defer gw.Close()
+		tw = tar.NewWriter(gw)
+
+	case "bzip2":
+		// Add support for bzip2 compression
+		bw, err := bzip2.NewWriter(buf, nil) // Pass nil as the second argument, since bzip2 required to two argument so
+		if err != nil {
+			return err
+		}
+		defer bw.Close()
+		tw = tar.NewWriter(bw)
+	default:
+		return fmt.Errorf("unsupported compression type: %s", *compression)
+
+	}
+	defer tw.Close()
+
+	var filesToArchive []string
+
+	// Iterate over collectMetrics and add them to the tar archive
+	for _, v := range collectMetrics {
+		for i := 0; i < int(v.fileCount); i++ {
+			metricName, err := getMetricName(v)
+			if err != nil {
+				log.Fatalf("createArchive: %v", err)
+			}
+			filename := fmt.Sprintf("%s.%05d", metricName, i)
+			filesToArchive = append(filesToArchive, filename)
+		}
+	}
+	//custom filename collect nd add them to the tar archiv
+	if *metric != "" {
+		for i := 0; i < int(customMatricCount); i++ {
+			filename := fmt.Sprintf("%s.%05d", *metric, i)
+			filesToArchive = append(filesToArchive, filename)
+		}
+	}
+
+	for _, file := range filesToArchive {
+		// Add the file to the tar archive
+		err := addToArchive(tw, file)
+		if err != nil {
+			return err
+		}
+		// Remove the file after it has been added to the archive
+		err = os.Remove(file)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
```

Comment:
Why use a loop to generate the list of the files, then iterate through the generated list? You could just write the files to the tarball inside the loop that's generating the list.

I recommend leaving the export files on disk until the tarball has been created successfully. If something goes wrong during tarball creation, the customer could still tar the files up manually. This isn't possible if they've been deleted after being added to the tarball.

You can use the existing `cleanFiles` func to do the cleanup at the end. It takes the file prefix and the file count, so it has to be called once for each enabled export.